### PR TITLE
Read and store gnu build id from loaded core file

### PR DESCRIPTION
**Pull request by GeorgeHuyubo**
As we have debuginfod as symbol locator available in lldb now, we want to make full use of it.
In case of post mortem debugging, we don't always have the main executable available.
However, the .note.gnu.build-id of the main executable(some other modules too), should be available in the core file, as those binaries are loaded in memory and dumped in the core file.

We try to iterate through the NT_FILE entries, read and store the gnu build id if possible. This will be very useful as this id is the unique key which is needed for querying the debuginfod server.

Test:
Build and run lldb. Breakpoint set to https://github.com/llvm/llvm-project/blob/main/lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp#L147
Verified after this commit, module_uuid is the correct gnu build id of the main executable which caused the crash(first in the NT_FILE entry)

Previous PR: https://github.com/llvm/llvm-project/pull/92078 was mistakenly merged. This PR is re-opening the commit.

### CODE REVIEW

File: lldb/include/lldb/Target/Process.h [hunk 1]
```diff
@@ -406,6 +406,36 @@ class Process : public std::enable_shared_from_this<Process>,
                                   lldb::StateType state);
   } Notifications;
 
+  class ProcessMemoryIterator {
```
Comment (labath) [inline lldb/include/lldb/Target/Process.h:409]: AFAICT, this isn't necessary anymore. If you think it's useful, you can make it a separate patch.

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp [hunk 2]
```diff
@@ -983,6 +1005,73 @@ llvm::Error ProcessElfCore::ParseThreadContextsFromNoteSegment(
...
+bool ProcessElfCore::IsElf(const lldb::addr_t address) {
+  uint8_t buf[4];
+  Status error;
+  size_t byte_read = ReadMemory(address, buf, 4, error);
+  if (byte_read != 4)
+    return false;
+  return elf::ELFHeader::MagicBytesMatch(buf);
+}
+
+std::optional<UUID> ProcessElfCore::FindNote(const lldb::addr_t address,
```
Comment (labath) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:1017]: Please give this a more specific name (to say what kind of note its searching for and where). An uninformed reader would think this is searching for a note in the core file itself.

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp [hunk 3]
```diff
@@ -983,6 +1005,73 @@ llvm::Error ProcessElfCore::ParseThreadContextsFromNoteSegment(
...
+  Status error;
+  size_t byte_read = ReadMemory(address, buf, elf_header_size, error);
+  DataExtractor data(buf, 4096, GetByteOrder(), addr_size);
+  lldb::offset_t offset = ehdr_phoff_offset;
+  lldb::offset_t phoff = data.GetAddress(&offset);
+
+  offset = ehdr_phentsize_offset;
+  lldb::offset_t phentsize = data.GetU16(&offset);
+  offset = ehdr_phnum_offset;
+  lldb::offset_t phnum = data.GetU16(&offset);
```
Comment (labath) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:1038]: Reuse the existing `ELFHeader` class for this. I.e., read the necessary amount of bytes (you can just always read `sizeof(Elf64_Ehdr)` as that's larger) into a DataExtractor and call `header.Parse(extractor, 0)`.

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp [hunk 4]
```diff
@@ -983,6 +1005,73 @@ llvm::Error ProcessElfCore::ParseThreadContextsFromNoteSegment(
...
+    if (byte_read != phentsize)
+      break;
+    offset = 0;
+    uint32_t p_type = data.GetU32(&offset);
+    if (p_type != llvm::ELF::PT_NOTE)
+      continue;
+    offset = ELFOFFSETOF(Phdr, p_vaddr);
+    lldb::addr_t p_vaddr = data.GetAddress(&offset);
+    offset = ELFOFFSETOF(Phdr, p_memsz);
+    lldb::addr_t p_memsz = data.GetAddress(&offset);
```
Comment (labath) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:1054]: Similarly, you can use `ELFProgramHeader::Parse` for this bit.

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp [hunk 5]
```diff
@@ -983,6 +1005,73 @@ llvm::Error ProcessElfCore::ParseThreadContextsFromNoteSegment(
...
+        offset < p_memsz &&
+        data.GetU32(&offset, &note, sizeof(Section_Note) / sizeof(uint32_t))) {
+      if (note.namesz == 4 && note.type == type) {
+        const char *name = data.GetCStr(&offset);
+        if (name && strcmp("GNU", name) == 0)
+          return UUID(
+              llvm::ArrayRef<uint8_t>(buf + offset, note.descsz /*byte size*/));
+      }
+      offset += note.namesz + note.descsz;
+    }
```
Comment (labath) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:1070]: And `parseSegment` for this (despite the generic name, this is really for parsing note segments, maybe you could rename it while you're in here).

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.h [hunk 6]
```diff
@@ -158,6 +165,16 @@ class ProcessElfCore : public lldb_private::PostMortemProcess {
   // Returns number of thread contexts stored in the core file
   uint32_t GetNumThreadContexts();
 
+  // Populate gnu uuid for each NT_FILE entry
+  void UpdateBuildIdForNTFileEntries();
+
+  // Returns the value of certain type of note of a given start address
+  std::optional<lldb_private::UUID> FindNote(const lldb::addr_t address,
```
Comment (labath) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.h:172]: `const` on a value parameter (pointer and reference parameters are of course different) declaration don't actually do anything, and I don't believe we usually use them.

If you really want to you can still make them `const` in [the function definition](https://godbolt.org/z/PxM739P84) even if they're not const in the header (though we usually don't do that either).

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp [hunk 7]
```diff
@@ -271,6 +282,17 @@ Status ProcessElfCore::DoLoadCore() {
   return error;
 }
 
+void ProcessElfCore::UpdateBuildIdForNTFileEntries() {
+  if (!m_nt_file_entries.empty()) {
```
Comment (labath) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:286]: Any specific reason doing this as a separate pass (and not from inside parsing loop on line 936)?

**Review by labath — COMMENTED** [review]
This looks much better, though, after looking at this closer, I don't think it's necessary to reimplement the elf parsing code. We already have all the necessary data structures for parsing notes in the core file itself, and it looks like it would be fairly easy to reuse them for this purpose as well.

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp (hunk 7 above)
Comment (kevinfrei) [reply lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:286]: I'd suggest adding it to that pass as well, though it will be a little bit messier, as NT_PRPSINFO and NT_GNU_BUILD_ID have the same values, so they depend on the Name field (CORE and GNU, respectively...). Do a check for the GNU name at the very top (line 902ish?) and handle NT_GNU_BUILD_ID's before the rest of that loop body, or dealing with PRPSINFO logic will get much messier...

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp (hunk 2 above)
Comment (kevinfrei) [reply lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:1017]: I'd suggest instead, implement support for the Note section in the same fashion as the other ELF components, with a "parse" method and the like (check out ThreadElfCore.h) and integrated it into the top level loop, same as the other Program Header structures.

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.h [hunk 8]
```diff
@@ -117,6 +117,13 @@ class ProcessElfCore : public lldb_private::PostMortemProcess {
     lldb::addr_t end;
     lldb::addr_t file_ofs;
     std::string path;
+    lldb_private::UUID uuid; //.note.gnu.build-id
+  };
+
+  struct Section_Note {
+    uint32_t namesz;
+    uint32_t descsz;
+    uint32_t type;
```
Comment (kevinfrei) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.h:126]: Move this to the ThreadElfCore.h header and add a 'parse' method to it, so the UUID is populated much like the PRPSINFO data and other useful info from the coredump.

**Review by kevinfrei — CHANGES_REQUESTED** [review]

File: lldb/include/lldb/Target/Process.h [hunk 9]
```diff
@@ -1649,6 +1679,26 @@ class Process : public std::enable_shared_from_this<Process>,
 
   lldb::addr_t ReadPointerFromMemory(lldb::addr_t vm_addr, Status &error);
 
+  /// Find a string within a memory region.
```
Comment (clayborg) [inline lldb/include/lldb/Target/Process.h:1682]: Change to:
```
/// Find bytes within a memory range.

File: lldb/include/lldb/Target/Process.h [hunk 10]
```diff
@@ -1649,6 +1679,26 @@ class Process : public std::enable_shared_from_this<Process>,
...
+  ///
+  /// This function searches for the string represented by the provided buffer
```
Comment (clayborg) [inline lldb/include/lldb/Target/Process.h:1684]: change to:
```
  /// This function searches for the bytes represented by the provided buffer
```

File: lldb/include/lldb/Target/Process.h [hunk 11]
```diff
@@ -1649,6 +1679,26 @@ class Process : public std::enable_shared_from_this<Process>,
...
+  /// within the memory range specified by the low and high addresses. It uses
+  /// a bad character heuristic to optimize the search process.
+  ///
+  /// \param[in] low The starting address of the memory region to be searched.
+  ///
+  /// \param[in] high The ending address of the memory region to be searched.
+  ///
+  /// \param[in] buffer A pointer to the buffer containing the string to be
+  /// searched.
```
Comment (clayborg) [inline lldb/include/lldb/Target/Process.h:1693]:
```
  /// \param[in] buffer A pointer to the buffer containing the bytes to be
  /// searched.
```

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp (hunk 2 above)
Comment (clayborg) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:1017]: Maybe the name of `FindNoteInCoreMemory`?

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp [hunk 12]
```diff
@@ -33,12 +35,17 @@
...
+#include "lldb/lldb-types.h"
 
 using namespace lldb_private;
 namespace ELF = llvm::ELF;
 
 LLDB_PLUGIN_DEFINE(ProcessElfCore)
 
+#define ELFOFFSETOF(T, M)                                                      \
+  addr_size == 4 ? offsetof(llvm::ELF::Elf32_##T, M)                           \
+                 : offsetof(llvm::ELF::Elf64_##T, M)
```
Comment (clayborg) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:47]: We should pass in `addr_size` here:
```
#define ELFOFFSETOF(type_name, member, addr_size) ...
```

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp [hunk 13]
```diff
@@ -983,6 +1005,73 @@ llvm::Error ProcessElfCore::ParseThreadContextsFromNoteSegment(
...
+}
+
+std::optional<UUID> ProcessElfCore::FindNote(const lldb::addr_t address,
+                                             const uint32_t type) {
+  if (!IsElf(address))
+    return std::nullopt;
+  const uint32_t addr_size = GetAddressByteSize();
+  const lldb::offset_t ehdr_phoff_offset = ELFOFFSETOF(Ehdr, e_phoff);
+  const lldb::offset_t ehdr_phentsize_offset = ELFOFFSETOF(Ehdr, e_phentsize);
+  const lldb::offset_t ehdr_phnum_offset = ELFOFFSETOF(Ehdr, e_phnum);
```
Comment (clayborg) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:1024]: pass `addr_size` to the macro

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp (hunk 3 above)
Comment (clayborg) [reply lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:1038]: header for the `ELFHeader` class is in: 
```
lldb/source/Plugins/ObjectFile/ELF/ELFHeader.h
```

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.h (hunk 8 above)
Comment (clayborg) [reply lldb/source/Plugins/Process/elf-core/ProcessElfCore.h:126]: or move it into ELFHeader.h and then add a parse method. We are using a ton of stuff already from ELFHeader.h, so we should follow that path

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp [hunk 14]
```diff
@@ -983,6 +1001,66 @@ llvm::Error ProcessElfCore::ParseThreadContextsFromNoteSegment(
...
+    if (byte_read != elf_header.e_phentsize)
+      break;
+    offset = 0;
+    elf::ELFProgramHeader program_header;
+    program_header.Parse(data, &offset);
+    if (program_header.p_type != llvm::ELF::PT_NOTE)
+      continue;
+
+    byte_read =
+        ReadMemory(program_header.p_vaddr, buf, program_header.p_memsz, error);
```
Comment (clayborg) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:1046]: Make a dynamic memory buffer to make sure we have enough bytes to store the note data as it could exceed 4096 bytes, and then read into this buffer:
```
std::vector<uint8_t> note_bytes;
note_bytes.resize(program_header.p_memsz);
byte_read =
        ReadMemory(note_bytes.data(), note_bytes.size(), program_header.p_memsz, error);
```

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp [hunk 15]
```diff
@@ -983,6 +1001,66 @@ llvm::Error ProcessElfCore::ParseThreadContextsFromNoteSegment(
...
+    if (byte_read != program_header.p_memsz)
+      continue;
+
+    auto notes_or_error = parseSegment(data, program_header.p_memsz);
```
Comment (clayborg) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:1050]: Create a DataBuffer for only the note data, and then we won't need to pass the `program_header.p_memsz` to `parseSegment(...)`:
```
DataExtractor note_data(note_bytes.data(), note_bytes.size(), GetByteOrder(), addr_size);
auto notes_or_error = parseSegment(note_data);
```

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.h [hunk 16]
```diff
@@ -117,6 +117,7 @@ class ProcessElfCore : public lldb_private::PostMortemProcess {
     lldb::addr_t end;
     lldb::addr_t file_ofs;
     std::string path;
+    lldb_private::UUID uuid; //.note.gnu.build-id
```
Comment (clayborg) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.h:120]: Add a comment letting the user know the UUID isn't actually in the NT_FILE_Entry:
```
// Add a UUID member for convenient access. The UUID value is not in the NT_FILE entries,
// we will find it in core memory and store it here for easy access.
```

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.h [hunk 17]
```diff
@@ -167,7 +178,8 @@ class ProcessElfCore : public lldb_private::PostMortemProcess {
   AddAddressRangeFromMemoryTagSegment(const elf::ELFProgramHeader &header);
 
   llvm::Expected<std::vector<lldb_private::CoreNote>>
-  parseSegment(const lldb_private::DataExtractor &segment);
+  parseSegment(const lldb_private::DataExtractor &segment,
+               unsigned long segment_size = 0);
```
Comment (clayborg) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.h:182]: revert

**Review by clayborg — CHANGES_REQUESTED** [review]

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp (hunk 7 above)
Comment (labath) [reply lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:280]: Makes sense. I though something like this might be the case.

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp [hunk 18]
```diff
@@ -983,6 +1001,66 @@ llvm::Error ProcessElfCore::ParseThreadContextsFromNoteSegment(
   }
 }
 
+bool ProcessElfCore::IsElf(lldb::addr_t address) {
+  uint8_t buf[4];
+  Status error;
+  size_t byte_read = ReadMemory(address, buf, 4, error);
+  if (byte_read != 4)
+    return false;
```
Comment (labath) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:1009]: I think you can just delete this and put the magic bytes check at on line ~1024, after you read the full header. Reading 4 bytes is not going to be appreciably faster than reading 64 (sizeof(Elf64_Ehdr)), and this will make the code look more like the one on lines 60--70 (which essentially does the same thing).

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp [hunk 19]
```diff
@@ -271,6 +276,17 @@ Status ProcessElfCore::DoLoadCore() {
   return error;
 }
 
+void ProcessElfCore::UpdateBuildIdForNTFileEntries() {
+  if (!m_nt_file_entries.empty()) {
+    for (NT_FILE_Entry &entry : m_nt_file_entries) {
+      std::optional<UUID> uuid =
```
Comment (labath) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:282]: Since UUID already has an IsValid() method I don't think wrapping it in an optional is not going to help (in fact, it will add ambiguity: Does returning an invalid UUID mean something different than std::nullopt?) Up to you, but I'd remove the optional here...

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp [hunk 20]
```diff
@@ -983,6 +1001,66 @@ llvm::Error ProcessElfCore::ParseThreadContextsFromNoteSegment(
...
+  uint8_t buf[4];
+  Status error;
+  size_t byte_read = ReadMemory(address, buf, 4, error);
+  if (byte_read != 4)
+    return false;
+  return elf::ELFHeader::MagicBytesMatch(buf);
+}
+
+std::optional<UUID> ProcessElfCore::FindNoteInCoreMemory(lldb::addr_t address,
+                                                         uint32_t type) {
```
Comment (labath) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:1014]: I don't think this `type` argument is useful, as this function already a UUID, and so it can't be used to parse any other kind of a note. I'd recommend inlining NT_GNU_BUILD_ID into the function (and renaming it to something like FindBuidIdInCoreMemory).

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp [hunk 21]
```diff
@@ -983,6 +994,67 @@ llvm::Error ProcessElfCore::ParseThreadContextsFromNoteSegment(
...
+      continue;
+
+    std::vector<uint8_t> note_bytes;
+    note_bytes.resize(program_header.p_memsz);
+
+    byte_read = ReadMemory(program_header.p_vaddr, note_bytes.data(),
+                           program_header.p_memsz, error);
+    if (byte_read != program_header.p_memsz)
+      continue;
+    assert(sizeof(buf) >= program_header.p_memsz);
```
Comment (labath) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:1040]: This looks like an obsolete assertion as the program headers are read into a dynamically sized vector now.

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp [hunk 22]
```diff
@@ -983,6 +994,67 @@ llvm::Error ProcessElfCore::ParseThreadContextsFromNoteSegment(
...
+                               GetByteOrder(), addr_size);
+    auto notes_or_error = parseSegment(segment_data);
+    if (!notes_or_error)
+      return invalid_uuid;
+    for (const CoreNote &note : *notes_or_error) {
+      if (note.info.n_namesz == 4 &&
+          note.info.n_type == llvm::ELF::NT_GNU_BUILD_ID) {
+        if ("GNU" == note.info.n_name)
+          return UUID(llvm::ArrayRef<uint8_t>(
+              note.data.GetDataStart(), note.info.n_descsz /*byte size*/));
```
Comment (labath) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:1051]: In case of a corrupted core file, `note.data` can contain less than `n_descsz` bytes, and this will create a invalid ArrayRef (and possibly crash lldb). Writing this like `return UUID(note.data.GetData())` will ensure that does not happen, although it could return a truncated UUID in that case. If you want to check that case explicitly, then I'd recommend something like:
```
if(note.data.ValidOffsetForDataOfSize(0, note.info.n_descsz))
  return UUID(note.data.GetData().take_front(note.info.n_descsz);
```

**Review by labath — COMMENTED** [review]
Very close. Just a couple of details.

BTW, I believe the recommended workflow for working with pull requests is to put your new changes as additional commits on top and then squash them all together after merging. The GitHub UI just works better that way.

**Review by labath — APPROVED** [review]
